daemons/attrd/attrd_commands.c

```diff
@@ -731,9 +732,18 @@ attrd_current_only_attribute_update(crm_node_t *peer, xmlNode *xml)
                 crm_trace("Syncing %s[%s] = %s to everyone.(from local only attributes)", a->id, v->nodename, v->current);
 
                 build = TRUE;
+                /* Attributes held only by own node with delay are forcibly written. */
                 build_attribute_xml(sync, a->id, a->set, a->uuid, a->timeout_ms, a->user, a->is_private,
-                            v->nodename, v->nodeid, v->current);
+                            v->nodename, v->nodeid, v->current, (a->timeout_ms && a->timer ? TRUE : FALSE));
             } else {
+                if (a->force_write) {
+                    /* After being reflected in Writer, clear forced writing flag and change flag are cleared. */
+
+                    crm_trace("Clear the forced write flag of %s attribute", a->id);     
+                    a->changed = FALSE;
+                    a->force_write = FALSE;
+                }
```

Comment:
I think this block might be unused and unnecessary.

attrd_current_only_attribute_update() is called only when receiving an ATTRD_OP_SYNC_RESPONSE from a new writer; if the (re-)joining node sends an ATTRD_OP_SYNC_RESPONSE with local values back to the writer, the writer will never get to this block, because it won't see the joining node as a new writer.

I don't think it's needed, since changed and force_write are cleared just before actually writing.

Reply:
Hi Ken,

Thanks for your comment.


The first ATTRD_OP_SYNC_RESPONSE exchanges are started at the time the divided attrds are merged.
At this time, attrd's new Writer has not been decided.

```
(snip)
void
attrd_peer_change_cb(enum crm_status_type kind, crm_node_t *peer, const void *data)
{
    bool remove_voter = FALSE;

    switch (kind) {
(snip)
        case crm_status_nstate:
            if (safe_str_eq(peer->state, CRM_NODE_MEMBER)) {
                /* If we're the writer, send new peers a list of all attributes
                 * (unless it's a remote node, which doesn't run its own attrd)
                 */
                if (attrd_election_won()
                    && !is_set(peer->flags, crm_remote_node)) {
                    attrd_peer_sync(peer, NULL);
                }

(snip)
```

After deciding in the election, the new Writer again sends ATTRD_OP_SYNC_RESPONSE with the forced flag synchronized content.

```
(snip)
gboolean
attrd_election_cb(gpointer user_data)
{
    attrd_declare_winner();

    /* Update the peers after an election */
    attrd_peer_sync(NULL, NULL);

    /* Update the CIB after an election */
    write_attributes(TRUE, FALSE);
    return FALSE;
}
(snip)
```

At this timing, I think that nodes that are not Writer should clear the forced flags that have been sent.


Best Regards,
Hideo Yamacuhi.

Reply:
Hi Hideo,

Thank you for the explanation. I overlooked that attrd_current_only_attribute_update() broadcasts its result, so all the nodes (not just the writer) will set a->force_write on any attributes updated by that function.

I think a better approach would be to set a->force_write in attrd_peer_update() only for the writer. Then the non-writers do not need to clear it. I don't think we need to clear a->changed in the non-writers; if for some reason the writer's write fails and a new writer takes over, it would be good for them to attempt to write it.

Reply:
Hi Ken,

Thanks for your comment.

>I think a better approach would be to set a->force_write in attrd_peer_update() only for the writer. Then the non-writers do not need to clear it. I don't think we need to clear a->changed in the non-writers; if for some reason the writer's write fails and a new writer takes over, it would be good for them to attempt to write it.

I will consider a fix a bit more about your opinion.

Many thanks,
Hideo Yamauchi.

Reply:
Hi Ken,

Regarding the a->changed flag, it is possible to modify my patch as follows so that only Writer sets and clears it.

```
(snip)
void 
attrd_current_only_attribute_update(crm_node_t *peer, xmlNode *xml)
{
(snip)
                if (a->force_write) {
                    /* After being reflected in Writer, clear forced writing flag and change flag are cleared. */

                    crm_trace("Clear the forced write flag of %s attribute", a->id);     
//Delete            a->changed = FALSE;
                    a->force_write = FALSE;
                }
(snip)
void
attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
(snip)

        if (is_force_write && a->timeout_ms && a->timer) {
            /* Save forced writing and set change flag. */
            /* The actual attribute is written by Writer after election. */
            crm_trace("Unchanged %s[%s] from %s is %s(Set the forced write flag)", attr, host, peer->uname, value);
//Delete    a->changed = TRUE;
            a->force_write = is_force_write;
        } else {
            crm_trace("Unchanged %s[%s] from %s is %s", attr, host, peer->uname, value);
        }
(snip)
void
write_attributes(bool all, bool ignore_delay)
{
    GHashTableIter iter;
    attribute_t *a = NULL;

    crm_debug("Writing out %s attributes", all? "all" : "changed");
    g_hash_table_iter_init(&iter, attributes);
    while (g_hash_table_iter_next(&iter, NULL, (gpointer *) & a)) {
        if (!all && a->unknown_peer_uuids) {
            // Try writing this attribute again, in case peer ID was learned
            a->changed = TRUE;
        } else if (a->force_write) {  //Add
            a->changed = TRUE;        //Add
        }

        if(all || a->changed) {
            /* When forced write flag is set, ignore delay. */
            write_attribute(a, (a->force_write ? TRUE : ignore_delay));
        } else {
            crm_debug("Skipping unchanged attribute %s", a->id);
        }
    }
}
(snip)
```


However, with regard to the a->force_write flag, since Writer is not decided when receiving the value, I think that it is impossible for me to set only Writer.

What do you think?

Best Regards,
Hideo Yamauchi.

Reply:
Hi Hideo,

If I'm following the code correctly, force_write can be set in an update only by attrd_current_only_attribute_update(), which can be called only when there is a new writer detected. So, I don't think the initial sync responses (when there is no writer) can get force_write set. Or am I missing some other way it can be set?

Reply:
Hi Ken,

I confirmed the behavior of attrd.

Actual attrd processing is as follows.

1. After the cluster is consolidated from fragmentation, each other node sends a sync_response message.
2. It receives a sync_response message and processes it with attrd_peer_message ().
3. Since attrd_check_for_ new_writer () and mutual nodes were Writer, peer_state is processed by election_won, so attrd_check_for_ new_writer () returns TRUE as a result.
4. Since attrd_check_for_new_writer () returns TRUE, it is executed when attrd_current_only_attribute_update () receives the first sync_response message.

As a result, Writer will be processed in a state that has not been decided yet.

Best Regards,
Hideo Yamauchi.

Reply:
Hi Hideo,

Thank you for taking the time to explain. I think I'm understanding a little better.

With step 4, each of the conflicting writers will broadcast updates for any of its own attributes that weren't known by the other writer. These updates will set F_ATTRD_IS_FORCE_WRITE appropriately. All other nodes will use this to set a->force_write.

Then, once a new election finishes, the winning writer will send another sync response with all its attributes. If this is the first message received from the new writer, peer_won will be true, and the other nodes will call attrd_current_only_attribute_update().

At this point, I am not sure the other nodes should clear a->force_write. They know the new writer has the attributes, but they do not know whether it has successfully written them to disk. If the writer fails to write to disk, and leaves the cluster, and one of the other nodes becomes the new writer, leaving a->force_write set may be safer (they will clear it once they write the attribute the first time).

Also, it might be possible for the new writer to send some other operation before sending its initial sync response, which means the other nodes won't see it as a new writer when getting the sync response, so this may not be the best place to clear it anyway.

Reply:
Hi Ken,

Thanks for your comment.

I certainly think you are right.
In conclusion, it is better not to clear a -> force_write.
What do you think?

If you agree, I will create a modified PR again.

Most importantly, what we care about is that this patch still does not work well.
As I wrote in the beginning, if the attraction of Election of attrd is delayed, we still can not solve the problem.
Do you have any good solutions to this case?

Best Regards,
Hideo Yamauchi.

Reply:
> I certainly think you are right.
> In conclusion, it is better not to clear a->force_write.
> What do you think?
> 
> If you agree, I will create a modified PR again.

Yes, I think that makes sense.
 
> Most importantly, what we care about is that this patch still does not work well.
> As I wrote in the beginning, if the attraction of Election of attrd is delayed, we still can not solve the problem.
> Do you have any good solutions to this case?

Unfortunately, no. It does look like transition-delay is the best current solution.

I could imagine something like this:

1. When attrd starts up, sees a new node join, or has an election in progress, it sets a new global variable to false.
2. After a new writer is elected and all attributes have been synchronized and successfully written to disk, the writer broadcasts a message that tells all attrd instances to set the global to true. (This is the most difficult part. I don't see a way to detect this reliably in a reasonable amount of time.)
3. When the DC sees a new node join, it queries the local attrd for the value of the global, passing the new node's name. If attrd doesn't recognize the node name as a peer, it returns false regardless of the actual global value (this ensures the DC doesn't get true before the attrd has seen the node). The DC repeats the query at intervals until the value is true (or a short timeout passes), then proceeds with the scheduler.

I think that would avoid the issue. But it would have to be resilient to nodes joining and/or leaving during the process, as well as the writer or DC changing during the process.

Reply:
Hi Ken,

Thanks for your comment.

> > I certainly think you are right.
> > In conclusion, it is better not to clear a->force_write.
> > What do you think?
> > If you agree, I will create a modified PR again.
> 
> Yes, I think that makes sense.

Okay!

> 
> > Most importantly, what we care about is that this patch still does not work well.
> > As I wrote in the beginning, if the attraction of Election of attrd is delayed, we still can not solve the problem.
> > Do you have any good solutions to this case?
> 
> Unfortunately, no. It does look like transition-delay is the best current solution.
> 
> I could imagine something like this:
> 
> 1. When attrd starts up, sees a new node join, or has an election in progress, it sets a new global variable to false.
> 2. After a new writer is elected and all attributes have been synchronized and successfully written to disk, the writer broadcasts a message that tells all attrd instances to set the global to true. (This is the most difficult part. I don't see a way to detect this reliably in a reasonable amount of time.)
> 3. When the DC sees a new node join, it queries the local attrd for the value of the global, passing the new node's name. If attrd doesn't recognize the node name as a peer, it returns false regardless of the actual global value (this ensures the DC doesn't get true before the attrd has seen the node). The DC repeats the query at intervals until the value is true (or a short timeout passes), then proceeds with the scheduler.
> 
> I think that would avoid the issue. But it would have to be resilient to nodes joining and/or leaving during the process, as well as the writer or DC changing during the process.

Thank you for valuable opinion.
There are certainly difficult places, but I think that it is a useful correction plan.
I would like to exchange further opinions with you on the corrections around this a little more.

Once this PR is closed.
Regarding this correction, I will issue a new PR so please confirm.

Many thanks!
Hideo Yamauchi.